### PR TITLE
[doc] Add Fault Tolerant Documentation Page

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

This PR adds documentation for Fault Tolerant Training.

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: docs/source/advanced/fault_tolerant_training.rst [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+        def __getitem__(self, index):
+            return self.data[index]
+
+        def __len__(self):
+            return self.len
+
+If you are using a single iterable-based dataset, there are some limitations.
+You need to use and expose a sampler within your dataset.
+
+This won't be supported as Lightning can't infer you are iterating over count.
```
[the remark this answers is not in the document]
Comment (awaelchli) [reply docs/source/advanced/fault_tolerant_training.rst]: LGTM!

File: docs/source/advanced/fault_tolerant_training.rst [hunk 2]
```diff
@@ -0,0 +1,155 @@
+Fault-tolerant Training
+=======================
+
+.. warning:: Fault-tolerant Training is currently an experimental feature within Lightning.
+
+Fault-tolerant Training is an internal mechanism that enables PyTorch Lightning to recover from a failure in hardware or software.
```
Comment (SeanNaren) [inline docs/source/advanced/fault_tolerant_training.rst:6]:
```suggestion
Fault-tolerant Training is an internal mechanism that enables PyTorch Lightning to recover from a hardware or software failure.
```

**Review by SeanNaren — APPROVED** [review]

**Comment by SeanNaren** [comment]
Extremely excited, great work @awaelchli @tchaton and others who helped in the effort to add this to Lightning :)

**Review by ethanwharris — APPROVED** [review]
LGTM :smiley:

**Pull request merged**